semgrep-core/src/metachecking/Check_rule.ml

```diff
@@ -80,66 +78,80 @@ let error env t s =
   Common.push err env.errors
 
 (*****************************************************************************)
-(* Formula *)
+(* Checks *)
 (*****************************************************************************)
 
-let equal_formula x y = AST_utils.with_structural_equal R.equal_formula x y
-
-let check_formula env (lang : Xlang.t) f =
-  (* check duplicated patterns, essentially:
-   *  $K: $PAT
-   *  ...
-   *  $K2: $PAT
-   * but at the same level!
-   *
-   * See also now semgrep-rules/meta/identical_pattern.sgrep :)
-   *)
-  let rec find_dupe f =
+let unknown_metavar_in_comparison env f =
+  let rec collect_metavars f : MV.mvar Set.t =
     match f with
-    | P _ -> ()
-    | Not (_, f) -> find_dupe f
-    | Or (t, xs)
-    | And (t, xs, _) ->
-        let rec aux xs =
-          match xs with
-          | [] -> ()
-          | x :: xs ->
-              (* todo: for Pat, we could also check if exist PatNot
-               * in which case intersection will always be empty
-               *)
-              xs
-              |> List.iter (fun y ->
-                     if equal_formula x y then
-                       let tx, ty = (R.tok_of_formula x, R.tok_of_formula y) in
-                       let kind = R.kind_of_formula x in
-                       error env ty
-                         (spf "Duplicate %s of %s at line %d" kind kind
-                            (PI.line_of_info tx)));
-              xs
-              |> List.iter (fun y ->
-                     if equal_formula (Not (t, x)) y then
-                       let tx, ty = (R.tok_of_formula x, R.tok_of_formula y) in
-                       let kind = R.kind_of_formula x in
-                       error env ty
-                         (spf "Unsatisfiable formula with %s at line %d" kind
-                            (PI.line_of_info tx)));
-              aux xs
+    | P ({ pat = _pat; pstr = pstr, _; pid = _pid }, _) ->
+        let words = Str.split (Str.regexp "[^a-zA-Z0-9_$]") pstr in
```

Comment:
This is pretty jank. Originally I traversed the tree and extracted metavariables, but I found that this wouldn't work with generic mode, and adding that (and other potential parsing modes we could add) seemed annoying. On the flip side, if we ever change our metavariable syntax, this is fragile. Not sure what the best approach is

Reply:
I'm not completely following but based on your comment, it would be reasonable to expect any matcher that returns bound metavariables to be able to tell us in advance which metavariables it would bind upon success. Generic mode uses metavariables similarly to semgrep. pattern-regex also does (or should) have the ability to bind metavariables to matched substrings ("groups" in regex jargon). In the future, we could have also other analyzers (the new `metavariable-analysis`) that could return a collection of bound metavariables. This collection should be known statically without resorting to special hacks.

A matcher (semgrep, generic, regex, ...) would have to follow a module signature that might look something like:
```ocaml
type target = ... (* file or a string *)

type search_results = ... (* list of matches with bound metavariables or errors *)

module type Matcher = sig
  type param
  type parsed_target
  val metavariables : param -> string list
  val parse : target -> parsed_target
  val search : param -> parsed_target -> search_results
end
```